src/Features/CSharp/Portable/ConvertSwitchStatementToExpression/ConvertSwitchStatementToExpressionDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Linq;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertSwitchStatementToExpression
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal sealed partial class ConvertSwitchStatementToExpressionDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        public ConvertSwitchStatementToExpressionDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.ConvertSwitchStatementToExpressionDiagnosticId,
+                new LocalizableResourceString(nameof(CSharpFeaturesResources.Convert_switch_statement_to_expression), CSharpFeaturesResources.ResourceManager, typeof(CSharpFeaturesResources)),
+                new LocalizableResourceString(nameof(CSharpFeaturesResources.Use_switch_expression), CSharpFeaturesResources.ResourceManager, typeof(CSharpFeaturesResources)))
+        {
+        }
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxNodeAction(AnalyzeSyntax, SyntaxKind.SwitchStatement);
+
+        private void AnalyzeSyntax(SyntaxNodeAnalysisContext context)
+        {
+            var switchStatement = context.Node;
+            if (switchStatement.GetDiagnostics().Any(d => d.Severity == DiagnosticSeverity.Error))
+            {
+                return;
+            }
+
+            if (!Analyzer.CanConvertToSwitchExpression(switchStatement))
+            {
+                return;
+            }
+
+            context.ReportDiagnostic(Diagnostic.Create(Descriptor,
+                // Report the diagnostic on the "switch" keyword.
+                switchStatement.GetFirstToken().GetLocation()));
+        }
+
+        public override bool OpenFileOnly(Workspace workspace)
+            => false;
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SyntaxAnalysis;
```

Comment:
It's unclear to me if this is allowed.  @mavasani Is this ok if he uses RegisterSyntaxNodeAction?  or only if he used RegisterSyntaxTreeAction